### ISSUE TITLE
Not yet using `tree-sitter.json`

### ISSUE THREAD

**Issue by sogaiu** [sogaiu/tree-sitter-clojure#64]
I think at some point we should start using `tree-sitter.json`.  The file is described [here](https://github.com/tree-sitter/tree-sitter/issues/3637).

Also, there is some more info about eventual removal of the `tree-sitter` section from `package.json` [here](https://github.com/tree-sitter/tree-sitter/pull/3700).

I'd like to get rid of `package.json`, but I think Cursorless may still depend on it, so that may not happen so soon.

---

(On a side note, the `build-wasm` command may be going away in 0.24; with recent releases you can already use `build --wasm` the replacement.  My attempts at getting the older subcommand to stay around longer did not meet with success.  Backward compatibility of the cli API does not rank that highly priority-wise AFAIU.)

**Comment by sogaiu** [comment]
At some point it looks like the `tree-sitter` cli's `generate` subcommand may result in the mutating of `package.json` and the creating of a `tree-sitter.json` file.

Some of the values that got filled in were wrong for our case (e.g. license, repository URL, and binding-generation):

```json

  "grammars": [
    {
      "name": "clojure",
      "camelcase": "Clojure",
      "scope": "source.clojure",
      "path": ".",
      "file-types": [
        "bb",
        "clj",
        "cljc",
        "cljs"
      ]
    }
  ],
  "metadata": {
    "version": "0.0.13",
    "license": "MIT",
    "description": "Clojure grammar for tree-sitter",
    "links": {
      "repository": "https://github.com/tree-sitter/tree-sitter-clojure"
    }
  },
  "bindings": {
    "c": true,
    "go": true,
    "node": true,
    "python": true,
    "rust": true,
    "swift": true
  }
}
```

It may be that pre-emptively creating a `tree-sitter.json` will leave our `package.json` file alone.

Below is a draft of a possible `tree-sitter.json`:

```json
{
  "grammars": [
    {
      "name": "clojure",
      "camelcase": "Clojure",
      "scope": "source.clojure",
      "path": ".",
      "file-types": [
        "bb",
        "clj",
        "cljc",
        "cljs"
      ]
    }
  ],
  "metadata": {
    "version": "0.0.13",
    "description": "Clojure grammar for tree-sitter"
  }
}
```

Unfortunately, it looks like version info is now duplicated across two files.  We may be stuck with this unless / until Cursorless stops needing `package.json`.

**Comment by sogaiu** [comment]
The situation of grammars missing `tree-sitter.json` files was mentioned in [a comment as part of a recent PR](https://github.com/tree-sitter/tree-sitter/pull/4135#issuecomment-2613785101):

> Ok, there's 3 out of 356 grammars with no tree-sitter.json file used in nvim-treesitter that must be generated from the grammar. The config file is sort of mandatory for this feature, as we use the version defined in this file. That means <1% of grammars are impacted, and I've gone ahead and put up PRs to migrate them 2 days ago now - if they don't bother to update, then that's on them, and they cannot use ABI 15 then (but nothing is broken with ABI 14).

**Comment by sogaiu** [comment]
This is a reminder to update the [what and why document](https://github.com/sogaiu/tree-sitter-clojure/blob/40c5fc2e2a0f511a802a82002553c5de00feeaf4/doc/what-and-why.md) if `tree-sitter.json` is added.

**Comment by george-gayton** [comment]
Selfishly I'd like to request the inclusion of `tree-sitter.json`.

My use case is that I'd like to add Clojure support to [this](https://github.com/cucumber/language-service) cucumber service. The service requires the `tree-sitter.json` format to build - although I haven't spent much time into looking into alternatives.

That cucumber service currently supports go, java, javascript, php, python, ruby, rust and typescript - which all use `tree-sitter.json` with duplicated data in `package.json`.

**Comment by sogaiu** [comment]
I think at this point you are better off forking and adding it yourself.

It's unlikely that this change will be made soon.

**Comment by NoahTheDuke** [comment]
what is hurt or lost by adding it?

**Comment by george-gayton** [comment]
Is the problem with your approach [here](https://github.com/sogaiu/tree-sitter-clojure/issues/64#issuecomment-2401732994) the duplicate information, in which case would a pre-commit hook handle this case?

**Comment by george-gayton** [comment]
For a use case that seems unable to proceed without official support: [convo](https://github.com/cucumber/language-service/pull/271#issuecomment-3421646227).

**Comment by sogaiu** [comment]
I don't follow the logic of that statement.

The [link in the linked comment](https://github.com/sogaiu/tree-sitter-clojure/blob/6e41628e9d18b19caea1cb1d72aae4ccff5bdfe9/doc/what-and-why.md#about-the-bindings) is not up-to-date, nor is this quoted content:

> The bindings are not tested in any way and only really exist as a side-effect of running the tree-sitter cli's generate subcommand.

That commit is from around 2023-03-17.  The bindings were removed in [this commit](https://github.com/sogaiu/tree-sitter-clojure/commit/97964b703622a15c46dd0a94a58253bdbc1315da) from around 2024-01-28.

At the time of this writing, [this](https://github.com/sogaiu/tree-sitter-clojure/blob/e43eff80d17cf34852dcd92ca5e6986d23a7040f/doc/what-and-why.md#about-bindings) is the latest, where it starts off:

> The repository does not host any bindings (e.g. for Rust, Node, or other programming language).
>
> They should be straight-forward to generate as long as one has a suitable tree-sitter cli and the grammar.js file mentioned above.

People are free to fork this repository and generate their own bindings.

I am not a node.js / npm community participant, nor do I plan on becoming one in the near future.  I don't believe it makes sense to expect a person in that kind of situation to create and maintain bindings in a responsible manner.

Please see [this comment](https://github.com/sogaiu/tree-sitter-janet-simple/issues/6#issuecomment-3374862655) in the issue of another grammar with a similar conversation.

**Comment by george-gayton** [comment]
Yeah this is understandable. I can see I've derailed this thread, apologies.

Thanks for the quick reply.

**Comment by wetneb** [comment]
I made [a fork with `tree-sitter.json` and bindings](https://codeberg.org/grammar-orchard/tree-sitter-clojure-orchard/), which are uploaded to crates.io and PyPI for now.

It doesn't store generated files such as `parser.c` inside the repository, so if your workflow relies on their existence, it's worth downloading the archive that contains generated files, published in each release: https://codeberg.org/grammar-orchard/tree-sitter-clojure-orchard/releases

If people are interested in using it in other ways, I'm open to adapting the publishing workflows of grammars stored in this Codeberg org to enable those uses. Co-maintainers are also welcome :)
